File: mesh_service_example/src/main/java/com/meshtastic/android/meshserviceexample/ble/BleForegroundService.java
package com.meshtastic.android.meshserviceexample.ble;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import com.meshtastic.android.meshserviceexample.R;

/**
 * A simple foreground service for handling long-running BLE operations.
 * Note: This is a basic implementation and is not currently used by MainActivity,
 * but it is provided for project completeness.
 */
public class BleForegroundService extends Service {
    private static final String TAG = "BleForegroundService";
    private static final String CHANNEL_ID = "BleForegroundServiceChannel";
    private static final int NOTIFICATION_ID = 1;

    @Override
    public void onCreate() {
        super.onCreate();
        Log.d(TAG, "Service created");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG, "Service started");
        createNotificationChannel();
        Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("Meshtastic HR Bridge")
                .setContentText("BLE service is running in the background.")
                .setSmallIcon(R.mipmap.ic_launcher) // Make sure you have this icon
                .build();
        startForeground(NOTIFICATION_ID, notification);

        // Perform background tasks here.
        // For this example, we just start the service and keep it running.

        return START_NOT_STICKY;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        // We don't provide binding, so return null
        return null;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "Service destroyed");
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel serviceChannel = new NotificationChannel(
                    CHANNEL_ID,
                    "BLE Foreground Service Channel",
                    NotificationManager.IMPORTANCE_DEFAULT
            );
            NotificationManager manager = getSystemService(NotificationManager.class);
            if (manager != null) {
                manager.createNotificationChannel(serviceChannel);
            }
        }
    }
}

File: mesh_service_example/src/main/java/com/meshtastic/android/meshserviceexample/ble/BleForegroundService.kt
package com.meshtastic.android.meshserviceexample.ble

import android.app.*
import android.content.Context
import android.content.Intent
import android.os.IBinder
import androidx.core.app.NotificationCompat
import com.meshtastic.android.meshserviceexample.R

class BleForegroundService : Service() {

    companion object {
        const val ACTION_START = "ble.service.START"
        const val ACTION_STOP  = "ble.service.STOP"
        private const val CH_ID = "ble_conn"
        private const val NOTI_ID = 1001

        fun start(context: Context) {
            val i = Intent(context, BleForegroundService::class.java).apply { action = ACTION_START }
            context.startForegroundService(i)
        }
        fun stop(context: Context) {
            val i = Intent(context, BleForegroundService::class.java).apply { action = ACTION_STOP }
            context.startService(i)
        }
    }

    override fun onCreate() {
        super.onCreate()
        createChannel()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_START -> {
                val noti = NotificationCompat.Builder(this, CH_ID)
                    .setContentTitle(getString(R.string.app_name))
                    .setContentText("BLE connected (foreground)")
                    .setSmallIcon(R.mipmap.ic_launcher)   // ← 用 app icon，避免資源缺失
                    .setOngoing(true)
                    .build()
                startForeground(NOTI_ID, noti)

                // TODO: 如要在背景維持連線/自動重連，可把 HrPlxBleClient 放這裡
            }
            ACTION_STOP -> {
                stopForeground(STOP_FOREGROUND_REMOVE)
                stopSelf()
            }
        }
        return START_STICKY
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onDestroy() {
        super.onDestroy()
        // TODO: 釋放 BLE 資源（client.disconnect()）
    }

    private fun createChannel() {
        val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        if (nm.getNotificationChannel(CH_ID) == null) {
            nm.createNotificationChannel(
                NotificationChannel(CH_ID, "BLE connection", NotificationManager.IMPORTANCE_LOW)
            )
        }
    }
}
